### ISSUE TITLE
Review of proposed API

### ISSUE THREAD

**Issue by ashcrow** [ashcrow/victims-api#1]
@jasinner / @abn,

I'd appreciate feedback on the new [proposed API](https://github.com/ashcrow/victims-api#proposed-endpoints) before I go much further implementing them :smile:.

**Comment by jasinner** [comment]
We'll need a way to submit new vulnerability reports. Ideally we'd be able to do it by submitting a CVE in combination with:
- HASH
- (Maven) coordinates
- A file archive (which gets computed to a hash server side)

**Comment by abn** [comment]
In terms of submission this should be enabled today via https://github.com/victims/victims-web#secured-api-access - something on par should be considered.

HASH submissions should be treated with care as the fingerprint algorithm could have changed. That said, I am not sure if we should accept artifacts without coordinates? Note that regeneration becomes hard, and complexities are greater if we are dealing with hash submissions. 

@ashcrow added some extra comments on the README. Can we put the content on some API doc service? Swagger/Apiary/readme.io etc? Might be a better way to hash out the APIs we need.

**Comment by ashcrow** [comment]
@abn yeah I'll have swagger enabled in the near future :smile: 

In terms of the secured-api-access I think we will drop the apikey/sig and just use basic auth due to feedback I've gotten from some users.

**Comment by ashcrow** [comment]
@jasinner ah, of course! Do you have an idea of how you'd like it to look or is cloning v2's upload process fine?

**Comment by abn** [comment]
@ashcrow I would say API key is way to go especially as submissions should really happen via clients. Was the feedback on the mailing list or in person? I'd prefer keys especially for automation scenarios too.

**Comment by jasinner** [comment]
I'd say the API key is OK, as long as provide examples of how to use it. I
found it challenging to get my head around it with the current API. I agree
that most submissions should come from automated clients.
I look forward to seeing the documentation generated by Swagger, as the
documentation for the old API was pretty slim. I'm still trying to figure
out how to use it.

**Comment by ashcrow** [comment]
@abn they were in person similar to the complaints @jasinner noted above. Not well documented and/or more code than expected to make API calls.

Though I'm OK with keeping API as they way to access any non read-only endpoints.

**Comment by ashcrow** [comment]
@jasinner proposed upload endpoint added: https://github.com/ashcrow/victims-api/blob/master/ProposedAPI.md#submit. Let me know what you think.

**Comment by jasinner** [comment]
@ashcrow It seems it might not be compatible with Java submissions. If it's Java we need:
- groupId
- artifactId
- versionId
  This is handled by using http parameters per group in v2. 
  The full list of requirements for submission groups can be found in the SUBMISSION_GROUPS value here:
  https://github.com/victims/victims-web/blob/master/src/victims_web/config.py

**Comment by ashcrow** [comment]
@jasinner how about the extra data that is different per group be part of the upload JSON? For a jar it could be something along the lines of:

``` json
{
    "groupId": "...",
    "artifactId": "..", 
    "versionId": "...",
}
```

while an egg would be:

``` json
{
    "package": "...",
}
```

Remind me, for a jar is versionId == version and artifactId == name? If so then it would just be groupId needed as extra data.

**Comment by jasinner** [comment]
@abn While we had this extra data in v2, I'm starting to wonder if it's really necessary. It only really matters that we have the file, and know which CVE it's vulnerable to. Why do we care about the library information? It could be inconsistent anyway, and sometimes, such as with many Red Hat libraries, they aren't found in Maven central. 
Should we just leave this information out of a file submission, to keep it simple? I think it's better that we keep non-file vulnerability reports in https://github.com/victims/victims-cve-db, and use a trusted repo such as maven central as in index of library version to file data.
Perhaps we an offer another endpoint in v3 which only accepts library information, and looks up the file in Maven Central, or pep?

**Comment by ashcrow** [comment]
@jasinner as in query out to a maven repo or pypy for metadata?

**Comment by jasinner** [comment]
@ashcrow Not for metadata, but for the actually bits. We might be able to offer another endpoint which takes library metadata then looks those in up in Maven repo, or pypy for the corresponding file to hash.

**Comment by jasinner** [comment]
Actually, re-reading Arun's earlier comments regarding re-hashing we should be capturing the metadata so that we can rehash a submission if necessary. @ashcrow therefore we should capture information in JSON body, as you suggested. I think it's better to keep artifactId for Java. versionId, or version, either is fine.

That being said, it's going to be useless if metadata is not correct per the index (Maven central, or pypy). So we should aim to verify the metadata at submission time. Perhaps we can flag that the library was not found in the index to an admin before approving?

**Comment by ashcrow** [comment]
@jasinner that sounds reasonable to me.
1. Check provided metadata against archive metadata (if present). Archive metadata has higher weighting.
2. Check highest weighted metadata against central source. If central source doesn't exist, flag as not centrally available and continue with highest weighted metadata.

Sound good?

**Comment by jasinner** [comment]
Hi Steven,

The extra data in upload JSON is good.

You are right version = versionId, and artifactId = name. However I feel
it's better to leave artifactId, and not call it name for Java if we can.

Thanks,
Jason

**Comment by jasinner** [comment]
Reviewing this discussion again I think we should have only 3 endpoints:
**submit:**
multipart/form file-upload
`/upload`
filters cves

**search:**
Returns a specific hash instance
`/search/{hash}`
takes a sha512sum and returns a list of matching CVEs

**deep search:**
`/deepsearch`
filter files
takes a lists of file hashes and returns a list of matching CVE where CVE files are a subset of provided filter (for scanning of fat jars)

**Comment by ashcrow** [comment]
@jasinner 

> Reviewing this discussion again I think we should have only 3 endpoints:

I'm in favour of this 😄. Much of v1 and v2 were over-engineered with features people didn't really use.

> **submit:**
> multipart/form file-upload
> ``/upload``
> filters cves


> **search:**
> Returns a specific hash instance
> ``/search/{hash}``
> takes a sha512sum and returns a list of matching CVEs

Due to the size of a sha512sum I think that we should have searches provided in the body rather than that URI. So instead of ``GET``ing an object matching the request you ``PUT`` (aka create) a search and get the results. The difference would be this:

```
GET https://victims/api/v3/search/759e398379edd761da0dfa40438e5b0d0706c000b4ad7e559e86862854fde59247b40a49f48d9813f082ec4e85668e4a862aaf09491f06a04e51f1a949790148
```

versus

```
PUT https://victims/api/v3/search
```

With the data being something like:
```json
{"hash": "759e398379edd761da0dfa40438e5b0d0706c000b4ad7e559e86862854fde59247b40a49f48d9813f082ec4e85668e4a862aaf09491f06a04e51f1a949790148"}
```

or 

```json
["759e398379edd761da0dfa40438e5b0d0706c000b4ad7e559e86862854fde59247b40a49f48d9813f082ec4e85668e4a862aaf09491f06a04e51f1a949790148"]
```

Response would then be:

```json
["cve-...", ...]
```

or

```json
{"cves": {"hash": "759e398379edd761da0dfa40438e5b0d0706c000b4ad7e559e86862854fde59247b40a49f48d9813f082ec4e85668e4a862aaf09491f06a04e51f1a949790148": ["cve-....", ...]}}
```

Thoughts?

> **deep search:**
> ``/deepsearch``
> filter files
> takes a lists of file hashes and returns a list of matching CVE where CVE files are a subset of provided filter (for scanning of fat jars)

I think I follow what you mean here, but do you mind providing an example request and response?

**Comment by jasinner** [comment]
For **Search,** I'd say a POST is preferable to PUT, and including a single hash in the body is the most elegant solution:
`POST https://victims/api/v3/search`
`{"hash": "759e398379edd761da0dfa40438e5b0d0706c000b4ad7e559e86862854fde59247b40a49f48d9813f082ec4e85668e4a862aaf09491f06a04e51f1a949790148"}`
and the response would be:
`["cve-...", ...]`

For **DeepSearch** an example request would be
`POST https://victims/api/v3/deepsearch`
`{
	"hashes": [{
			"hash": "759e398379edd761da0dfa40438e5b0d0706c000b4ad7e559e86862854fde59247b40a49f48d9813f082ec4e85668e4a862aaf09491f06a04e51f1a949790148"
		},
		{
			"hash": "98711e7069a21c86139e02d17bb6faad284dcace423deaae20cf6f942ec5c6d507c26b4db9596fa6f02f9c7d8428ac9b9d4c465fcdbbf45ab3cad25ea91af915"
		}
	]
}`
and the response would be:

`["cve-...", ...]`

**Comment by ashcrow** [comment]
@jasinner makes sense to me. I have a working prototype for ``search`` and ``deepsearch``.

For ``upload`` are you thinking people will upload package files to be scanned? Or are they uploading files for inclusion?

**Comment by ashcrow** [comment]
I've added the code that matches with the last few comments here. PTAL @jasinner.

**Comment by jasinner** [comment]
Upload files for inclusion. The server would hash them (using java-hashing service for example) then add them to the database.
Taking a hash for scanning would take place on the client side, then compared using the hash via the API.
